### ISSUE TITLE
Generators - analyzer using references causing problems

### ISSUE THREAD

**Issue by mgravell** [dotnet/roslyn#43903]
I have an idea for an analyzer I'd like to add, reusing functionality that is already packaged in a netstandard2.0 package.

Repro is here: https://github.com/protobuf-net/protobuf-net.Grpc/pull/85/

In [my analyzer csproj](https://github.com/protobuf-net/protobuf-net.Grpc/pull/85/files#diff-cb786d2a6685de19379b4eccc11ef2ac) I have the two expected refs, plus a commented out one for the tool I want to use:

``` xml
	<ItemGroup>
		<PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="3.6.0-3.20207.2" PrivateAssets="all" />
		<PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.0.0-beta2.final" PrivateAssets="all" />
		<!--<PackageReference Include="protobuf-net.Reflection" Version="3.0.0-alpha.143" PrivateAssets="all" />-->
	</ItemGroup>
```

The analyzer currently just lists the additional files (I have a `<AdditionalFiles Include="test.proto" />`). With the above reference commented out, everything kinda works; however, if I uncomment the `<PackageReference>` **and do nothing else**, it stops working completely.

This may simply be a packaging thing, but... it isn't obvious how to fix this; I've tried numerous things - with/without the `PrivateAssets` hint, with dll references vs project references to the analyzer, etc.

Any guidance @chsienki ?

**Comment by jmarolf** [comment]
unfortunately nuget does not know how to pack analyzer dependencies and we've never done to work to call nuget to dynamically resolve analyzer dependencies at runtime. Effectively, you need to place the `protobuf-net.Reflection` contents next to your analyzer/source generator.

**Comment by chsienki** [comment]
Sounds like @jmarolf can guess what the error is. We have work to do to actually surface the errors, for now they're just swallowed unless you have a debugger on it.

Current 'best guidance' (very loosely speaking) is to `Debugger.Launch()` in your initialize method. Then you can attach and see what's actually failing during a compilation.

**Comment by mgravell** [comment]
@jmarolf I'm not entirely sure what that means in practice as either an author or consumer. When you say nuget doesn't know... so: how does one most correctly ship an analyzer? Is this:

1. don't have non-inbuilt dependencies, they won't work; inline everything into one assembly
2. If it can resolve one level; write your own nuspec by hand that ships everything rather than using transitive trees
3. Some exotic csproj flag I haven't used when packaging
4. Something else I haven't thought of?

Or am I misunderstanding the problem?

**Comment by davidfowl** [comment]
I assume 1.

**Comment by sharwell** [comment]
If your analyzer assembly has _unique_ dependencies (i.e. no other part of the hosting application references the dependency, and no other analyzers reference the dependency), you can place the dependency assembly in the same folder of the NuGet package where the analyzer assembly is.

If the dependency is not unique, this may or may not work, and failures may be non-deterministic. The two cases where I know this has occurred in the past both ended up treating the situation like (1): remove or inline the dependency.

**Comment by davidfowl** [comment]
Once you "do plugins" you have to solve all the same problems we solve today at build time with nuget (unification etc)

**Comment by jmarolf** [comment]
yeah, we haven't tackled runtime-dependency-resolution for analyzers. As @davidfowl says, solving this would likely need to start with additions to nuget.

**Comment by mgravell** [comment]
Oof. Right, I guess I'm going to have to forget about transitive package refs for now, and do something horrible in my build, either with some "and include these .cs files", or maybe some IL-merge/fody fun. This could get interesting.

**Comment by danielcrenna** [comment]
Was about to submit a repro for this because it seems counter-intuitive that you can't simply reference some other otherwise uncontroversial `netstandard2.0` assembly in your source generator library, without it failing with an obscure warning.

I guess the least horrible thing to do here would be to post-build copy the package references to the analyzer build folder?

**Comment by danielcrenna** [comment]
It doesn't look like copying the offending dependency into the same location as the analyzer .DLL itself is working for me, maybe because `System.Text.Json` is including the ASP.NET Core `FrameworkReference`

Example code:
```
<ItemGroup>
    <PackageReference Include="System.Text.Json" Version="5.0.0-preview.3.20214.6" PrivateAssets="All" GeneratePathProperty="true" />
</ItemGroup>
<Target Name="CopyPrivateReferences" AfterTargets="Build">
    <Exec Command="echo f | xcopy $(PkgSystem_Text_Json)\lib\$(TargetFramework)\System.Text.Json.dll $(TargetDir)\System.Text.Json.dll /Y"/>
</Target>
```

**Comment by danielcrenna** [comment]
And trying to ILMerge that single dependency quickly descended into madness:

```
<ItemGroup>
    <PackageReference Include="System.Text.Json" Version="5.0.0-preview.3.20214.6" PrivateAssets="All" GeneratePathProperty="true" />
    <PackageReference Include="System.Memory" Version="4.5.4" GeneratePathProperty="true" />
    <PackageReference Include="System.Buffers" Version="4.5.1" GeneratePathProperty="true" />
    <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="5.0.0-preview.3.20214.6" GeneratePathProperty="true" />
    <PackageReference Include="System.Text.Encodings.Web" Version="5.0.0-preview.3.20214.6" GeneratePathProperty="true"  />
    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.4" GeneratePathProperty="true" />
    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1"  GeneratePathProperty="true" />
</ItemGroup>
<Target Name="ILMerge" AfterTargets="Build">
    <Exec Command="$(ILMergeConsolePath) /out:$(AssemblyName).dll ^
          /lib:$(NuGetPackageRoot)NETStandard.Library\2.0.1\build\netstandard2.0\ref ^
          /lib:$(PkgSystem_Runtime_CompilerServices_Unsafe)\lib\$(TargetFramework) ^
          /lib:$(PkgSystem_Buffers)\lib\$(TargetFramework) ^
          /lib:$(PkgSystem_Memory)\lib\$(TargetFramework) ^
          /lib:$(PkgSystem_Text_Encodings_Web)\lib\$(TargetFramework) ^
          /lib:$(PkgSystem_Threading_Tasks_Extensions)\lib\$(TargetFramework) ^
          $(PkgSystem_Text_Json)\lib\$(TargetFramework)\System.Text.Json.dll ^
          $(PkgMicrosoft_Bcl_AsyncInterfaces)\lib\$(TargetFramework)\Microsoft.Bcl.AsyncInterfaces.dll ^
          "/>
</Target>
```

**Comment by tmat** [comment]
Whatever you do, please don't use `ILMerge`.

**Comment by tmat** [comment]
I think the only sane solution is for Roslyn to run analyzers and code generators on .NET Core.

**Comment by mgravell** [comment]
@tmat right now I'm thinking my best option is to hack a build that imports all the code I need directly as code, but: somewhere in this there is a "the story of package dependencies in analyzers is bad"

**Comment by danielcrenna** [comment]
@tmat Would love not to; how would you suggest I reference `System.Text.Json` from my source generator? It's ILMerge or copy-pasting the code off of GitHub. Neither one is good.

**Comment by jmarolf** [comment]
Currently working on a design to resolve this. Agree it needs to be solved

**Comment by danielcrenna** [comment]
> Currently working on a design to resolve this. Agree it needs to be solved

Ah, good concept. I didn't think of stepping back into Roslyn to create snapshot tests, and wasn't aware of `GeneratorDriver`. Many thanks.

**Comment by wieslawsoltes** [comment]
I am using it like this:
```XML
  <ItemGroup>
    <!-- Take a private dependency on Newtonsoft.Json (PrivateAssets=all) Consumers of this generator will not reference it.
         Set GeneratePathProperty=true so we can reference the binaries via the PKGNewtonsoft_Json property -->
    <PackageReference Include="Newtonsoft.Json" Version="12.0.1" PrivateAssets="all" GeneratePathProperty="true" />

    <!-- Package the generator in the analyzer directory of the nuget package -->
    <None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />

    <!-- Package the Newtonsoft.Json dependency alongside the generator assembly -->
    <None Include="$(PkgNewtonsoft_Json)\lib\netstandard2.0\*.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
  </ItemGroup>
```
as suggested in https://github.com/dotnet/roslyn/blob/master/docs/features/source-generators.cookbook.md#use-functionality-from-nuget-packages

But it is hard to debug when referencing source generator as nuget package, still getting errors `'System.IO.FileNotFoundException: Could not load file or assembly`, it works when using project reference.

**Comment by wieslawsoltes** [comment]
> I am using it like this:
> 
> ```
>   <ItemGroup>
>     
>     <PackageReference Include="Newtonsoft.Json" Version="12.0.1" PrivateAssets="all" GeneratePathProperty="true" />
> 
>     
>     <None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
> 
>     
>     <None Include="$(PkgNewtonsoft_Json)\lib\netstandard2.0\*.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
>   </ItemGroup>
> ```
> 
> as suggested in https://github.com/dotnet/roslyn/blob/master/docs/features/source-generators.cookbook.md#use-functionality-from-nuget-packages

And I had include all nuget PackageReference as dll's  in `analyzers/dotnet/cs`.

**Comment by kjkrum** [comment]
> Currently working on a design to resolve this. Agree it needs to be solved

Any movement on this?

**Comment by sharwell** [comment]
@wieslawsoltes Note that Newtonsoft.Json is one of the dependencies we were never able to get to work reliably.

**Comment by kjkrum** [comment]
@sharwell The dependency conflicts issue is a deal breaker. Not only would it be annoying and fragile to manually add and maintain the recursive dependencies of a library like NSwag, but even if it works for me, I can't guarantee that another user of my generator will not have a conflict with some other analyzer. As currently designed, I don't see non-trivial source generators ever being practical.

**Comment by jmarolf** [comment]
@kjkrum the only way around this (which is what you need to do for Visual Studio extensions and other .NET extensibility systems) is strong name sign your assemblies. Unless you are willing to do that we cannot guarantee the runtime will not fail to load some mis-matched version from another assembly.

**Comment by jmarolf** [comment]
> > Currently working on a design to resolve this. Agree it needs to be solved
> 
> @jmarolf Any movement on this?

I'll put up a proposal next week in dotnet/designs

**Comment by kjkrum** [comment]
@jmarolf Source generators that take the form of console applications packaged with build targets do not need to be strong name signed and AFAIK do not suffer from dependency conflicts. Those remain my go-to because [the issue with zombie MSBuild tasks locking files](https://developercommunity.visualstudio.com/t/the-file-is-locked-by-net-core-host/793032) was either never actually fixed, or was fixed after I gave up hoping for a fix, which was some time after it was claimed to be fixed. I'll watch for your proposal.

**Comment by jmarolf** [comment]
> Source generators that take the form of console applications packaged with build targets do not need to be strong name signed and AFAIK do not suffer from dependency conflicts.

Right, if you do the entire build again (in order to create the compilation context) then you do not need to load the assemblies into the same process at all. They can all live in separate processed. Unfortunately this kills build performance (this build twice model was [already possible](https://github.com/AArnott/CodeGeneration.Roslyn) without a roslyn api). If a console app called from a build target meets your need today I do not think you should move to the roslyn source generators API. The only advantage the roslyn API gives is being able to introspect into the compilation in order to make decisions about what should be generated.

We could have separate processes for all source generators in order to isolate them. This would either require us to rebuild the compiler to allow all of its internal state to be serializable (unclear what the performance implications would be here), or have each driver for source generators re-create the compilation context massively degrading build performance. 

I currently do not think this large re-write is worth it for the use case source generators were intended for: code generated in conjunction with the user typing. Especially since its unclear to me how often conflicting dependencies are going to happen in the wild. For now I only plan to have us show a useful diagnostic so that if this situation ever happens its clear to the developer what is going on.

**Comment by TorreyGarland** [comment]
> ```
> <ItemGroup>
>     
>     <PackageReference Include="Newtonsoft.Json" Version="12.0.1" PrivateAssets="all" GeneratePathProperty="true" />
> 
>     
>     <None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
> 
>     
>     <None Include="$(PkgNewtonsoft_Json)\lib\netstandard2.0\*.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
>   </ItemGroup>
> ```

using VS 2022 v 17.0.4 with this no longer works:

```
2>CSC : warning CS8785: Generator 'IncrementalGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'FileNotFoundException' with message 'Could not load file or assembly 'Newtonsoft.Json, Version=12.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The system cannot find the file specified.'

```

**Comment by pCYSl5EDgo** [comment]
Me too.

**Comment by pluraltouch** [comment]
It gets weird... 

I am completely rebuilding the whole solution which produces the CS8787. The warning is there, with the explicit assembly name,  source generator does not produce the output (as expected :-(   ), but monitoring all processes with procmon64 no process seems to be even looking for that particular .dll. (No surprise it does not find it) 

if I monitor for the sourcegenerator packages I can see VBCSCompiler is opening the dll from ..nuger\packages\...........\analyzers\dotnet\cs\.... As a crosscheck I monitor for only 'analyzers\dotnet\cs' path fragment, I see only the sourcegenerator package is searched (and successfully found there)

So having the referenced assembly there is completely obsolete, and it is a puzzler why I does not see with procmon64 any process to looking for the "missing" assembly...

**Comment by sharwell** [comment]
> Take a private dependency on Newtonsoft.Json

StyleCop Analyzers never found a way to make Newtonsoft.Json a reliable dependency of an analyzer (or source generator). We ended up embedding a different JSON library as source.

**Comment by pluraltouch** [comment]
> 

this issue is more generic and not about how to use json in a source generator...

**Comment by mgravell** [comment]
fair enough... hopefully this second version makes things more generic...

~StyleCop Analyzers~ *protobuf-net.BuildTools* never found a way to make ~Newtonsoft.Json~ *anything* a reliable dependency of an analyzer (or source generator). We ended up embedding ~a different JSON library~ *everything we needed* as source.

**Comment by michael-hawker** [comment]
I've been facing this same issue with a similar CS8784 warning this past week as well trying to include `YamlDotNet`.

It seems to work in a stand-alone setup with the suggested setup like this:
```xml
     <GetTargetPathDependsOn>$(GetTargetPathDependsOn);GetDependencyTargetPaths</GetTargetPathDependsOn>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.0.1" Pack="false" PrivateAssets="all" />
    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.3" Pack="false" PrivateAssets="all" />
    <PackageReference Include="YamlDotNet" Version="11.2.1" PrivateAssets="all" Pack="false" GeneratePathProperty="true" />
  </ItemGroup>

  <Target Name="GetDependencyTargetPaths">
    <ItemGroup>
      <TargetPathWithTargetPlatformMoniker Include="$(PKGYamlDotNet)\lib\netstandard1.3\YamlDotNet.dll" IncludeRuntimeDependency="false" />
    </ItemGroup>
  </Target>
```

But is failing when I try and use this setup in our larger project.

**Comment by WoLfulus** [comment]
This is still a thing apparently. The suggestions here haven't worked for me (referencing the analyzer in the same solution).

**Comment by KieranFoot** [comment]
I am using the below solution in my project and it's working in the current VS2022 build (17.4.3).

Here are the relevant parts that include the relevant files for debug and packaging.

```cs
	<ItemGroup>
		<PackageReference Include="ObjectDumper.NET" Version="3.5.2" PrivateAssets="all" pack="false" GeneratePathProperty="true"/>
	</ItemGroup>

	<ItemGroup>
		<None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
		<None Include="$(PkgObjectDumper_NET)\lib\netstandard2.0\*.dll" Pack="true" PackagePath="lib/netstandard2.0" Visible="false" />
	</ItemGroup>

	<Target Name="CopyObjectDumper" BeforeTargets="Build">
		<ItemGroup>
			<ObjectDumperFiles Include="$(PkgObjectDumper_NET)\lib\netstandard2.0\*.dll" />
		</ItemGroup>
		<Copy SourceFiles="@(ObjectDumperFiles)" DestinationFolder="$(OutputPath)" />
	</Target>
```

**Comment by kjkrum** [comment]
Generators need to run out of process. Performance takes a back seat to actually working and being useful.

**Comment by sharwell** [comment]
@kjkrum The performance penalty associated with process isolation for source generators isn't viable. Eventually we'll be able to separate them using AssemblyLoadContext in .NET Core, but that will only work once the remaining tools using .NET Framework move over.

**Comment by kjkrum** [comment]
@sharwell Would it be practical to make it optional and let users decide if it's viable?

**Comment by kkukshtel** [comment]
Was this ever resolved? I'm also looking at figuring out how to just add a System.Text.Json reference to my generator and unclear on how it's meant to work.

**Comment by limeniye** [comment]
The same problem with my simple generator:

[SourceGenerators.csproj](https://github.com/limeniye/uimarkup/blob/feat/shared-protocol-generation/src/frameworks/UIMarkup.Protocol.SourceGenerators/UIMarkup.Protocol.SourceGenerators.csproj)
[SomeLibraryNetStandard.csproj](https://github.com/limeniye/uimarkup/blob/feat/shared-protocol-generation/src/frameworks/UIMarkup.Protocol.Standard/UIMarkup.Protocol.Standard.csproj)

The error exception:
```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	CS8785	Generator 'ControlGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'InvalidOperationException' with message 'Collection was modified; enumeration operation may not execute.'.	UIMarkup.Protocol.Standard	R:\uimarkup\src\frameworks\UIMarkup.Protocol.Standard\CSC	1	Active	
```

But the files I'm expecting are being generated correctly.

**Comment by limeniye** [comment]
I resolved this issue by removing ```<TreatWarningsAsErrors>True</TreatWarningsAsErrors>``` from the **Directory.Build.props** file

**Comment by kjkrum** [comment]
@jmarolf If the "performance penalty" of console app source generators is viable (which it is) then running incremental source generators with process isolation is also viable. Do not make performance decisions for users. I'd rather wait minutes for a build than years for a minimum viable product.

**Comment by kjkrum** [comment]
@sharwell If the "performance penalty" of console app source generators is viable (which it is) then running incremental source generators with process isolation is also viable. Do not make performance decisions for users. I'd rather wait minutes for a build than years for a minimum viable product.

**Comment by CyrusNajmabadi** [comment]
@kjkrum Just an FYI that Jonathan has sadly passed away.  We're continuing to look at this space.  But it is frought with challenges.

**Comment by CyrusNajmabadi** [comment]
>  than years for a minimum viable product.

We are continuing to work in this space.  But we're a small team with a ton of responsibilities.  I do have it as a goal to make it so that:

1. people can just rebuild their SGs and have them picked up.
2. SGs can run in a moden core environment, not just standard2
3. the IDE experience can be entirely insulated from poorly behaved SGs (the norm unfortunately).

However, i would not expect this to come soon as each of these is a substantial amount of work, and we are often constrained by global requirements around compatibility.

**Comment by kjkrum** [comment]
@CyrusNajmabadi Sorry to hear about Jonathan. FWIW, I don't blame your team at all. I blame Microsoft for releasing a product with such incredible potential in a half-finished state, and then leaving its completion to a small team with a ton of responsibilities.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi  FWIW, I don't blame your team at all. I blame Microsoft for releasing a product with such incredible potential in a half-finished state, and then leaving its completion to a small team with a ton of responsibilities.

For every person who wants us waiting before releasing to get all these things done, we have just as many asking to release early in whatever state things are in just so they can use it :-)

That very thing happened today already :-D

**Comment by kjkrum** [comment]
@CyrusNajmabadi The early release is less of a problem than seemingly abandoning it for years afterward. My impression is that someone in management is under the mistaken impression that it's "done" and has assigned resources elsewhere.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi The early release is less of a problem than seemingly abandoning it for years afterward. My assumption is that someone in management is under the mistaken impression that it's "done" and has assigned resources elsewhere.

It has not been abandoned. Several people, including myself are working in this space. However, the costs are enormous here. Especially given the dramatic impact SGs have on every single other layer that sits above them.

**Comment by sharwell** [comment]
> If the "performance penalty" of console app source generators is viable (which it is) then running incremental source generators with process isolation is also viable. Do not make performance decisions for users.

There are different ways that we can end up making performance decisions for users, and not all of them are bad. For example:

* "We shouldn't expose this API because a user might accidentally cause performance to be bad." This is a fairly common theme in a product with an extensibility layer as rich as Visual Studio, but I always fight back hard against making decisions in this manner.
* "We shouldn't expose this API because it will be impossible for any user (internal or external), even with deep understanding of the performance of the entire system, to use the API in even the simplest form without catastrophic performance impact." When this situation occurs, it makes no sense to proceed in that direction since it gives a mistaken appearance of a new feature when in reality the feature cannot be used.

To date, running generators in an isolated process runs afoul of the _second_ case. In other words, we are not withholding a feature here; the requested feature simply does not exist.

**Comment by ByronMayne** [comment]
I ended up creating an open source solution [SourceGenerator.Foundations.](https://github.com/ByronMayne/SourceGenerator.Foundations) It figures out your NuGet packages and transitive references and embeds them as resources into your source generator. It then injects an assembly resolver for you.  In addition it wraps all the generators and captures all exceptions and logs them out to a custom console window in Visual Studio using EnvDTE.  No more of that "this failed to load" with no context what so ever.

**Comment by kjkrum** [comment]
I took @ByronMayne's solution to the minimalist extreme: [Dependency Magic](https://gitlab.com/kjkrum/dependency-magic). It's not a source generator and doesn't require any changes to your source generator.
